### ISSUE TITLE
Generic/inheritable config values

### ISSUE THREAD

**Issue by metaperl** [reahl/reahl#227]
I currently have 4 folders with configuration data:
```
  c:/appserv/check_recon/stage:
  total used in directory 28 available 164660596
  drwxrwxrwx  1 ogh Dmn Users 4096 11-12 14:01 0-local
  drwxrwxrwx  1 ogh Dmn Users 4096 11-27 12:18 1-dev
  drwxrwxrwx  1 ogh Dmn Users 4096 11-26 16:27 2-staging
  drwxrwxrwx  1 ogh Dmn Users 4096 11-12 14:01 3-qa
  drwxrwxrwx  1 ogh Dmn Users 4096 11-26 16:28 4-prod

  c:/appserv/check_recon/stage/0-local:
  total used in directory 6 available 164660596
  -rw-rw-rw-  1 ogh Dmn Users   58 11-12 13:05 mailutil.config.py
  -rw-rw-rw-  1 ogh Dmn Users  636 11-07 08:19 reahl.config.py
  -rw-rw-rw-  1 ogh Dmn Users  212 10-29 12:19 web.config.py

  c:/appserv/check_recon/stage/1-dev:
  total used in directory 7 available 164660596
  -rw-rw-rw-  1 ogh Dmn Users   58 11-12 13:05 mailutil.config.py
  -rw-rw-rw-  1 ogh Dmn Users  811 11-16 09:20 reahl.config.py
  -rw-rw-rw-  1 ogh Dmn Users  442 11-27 12:18 systemaccountmodel.config.py
  -rw-rw-rw-  1 ogh Dmn Users  271 10-30 16:56 web.config.py
```

However, many values should actually be specified once and used across folders. Is there any support for generic configuration which is only refined in a particular configuration folder if it is really needed?

**Comment by IwanVosloo** [comment]
Nope, there is no such support... we really only know about the single config directory.

I'd imagine that you could write Python code in your config that does something like that. Although I would be hesitant to depend too much on code in config files. They really should just be setting values to keys. We still need more experience with having decided to use a language like Python for config. I'd be conservative with that.

You can also use tools like [diff](https://en.wikipedia.org/wiki/Diff) and [patch](https://en.wikipedia.org/wiki/Patch_(Unix)) (or a script that does the same job) to create a config that is derived from another from that other + a patch file.

**Comment by metaperl** [comment]
>  I would be hesitant to depend too much on code in config files. They really should just be setting values to keys.

Not really, for instance, certain things are constant throughout staging levels and should be written once and used throughout levels. Other things are very specific on which staging level you are in.

The [traitlets design objectives](https://traitlets.readthedocs.io/en/stable/config.html#design-requirements) are very much in line with how I'd like to see Reahl config files setup.

**Comment by IwanVosloo** [comment]
Ok. (.. and traitlets look interesting...)

I guess I am just mentally revisiting our decision to go for executable config files in the light of [criticisms others have against it, such as that executable files are difficult for programs to edit](https://softwareengineering.stackexchange.com/questions/351126/how-bad-of-an-idea-is-it-to-use-python-files-as-configuration-files).

That said, I think we are in good company in the Python world in doing this...

I guess what we could to is to change commands that take a single config directory to take an ordered list of config directories. Instead of evaluating only expected files in one directory, we can run through the directories in order.

**Comment by IwanVosloo** [comment]
Here is a practical idea of what you can do now:

Assuming the following somewhere you can import it from (say, a module myconfigthings):

```
from reahl.component.config import ConfigAsDict
import os

def read_from_directory(config, other_directory):
    other_file = os.path.join(os.getcwd(), other_directory, config.filename)
    locals_dict = ConfigAsDict(config)
    locals_dict[config.config_key] = config
    with open(other_file) as f:
        exec(compile(f.read(), other_file, 'exec'), globals(), locals_dict)
```

You can, for example write web.config.py as follows:

```
from myconfigthings import read_from_directory

read_from_directory(web, 'base')

web.session_key_name = 'an_overriding_value'

```

This assumes the two config directories to be next to each other in the current directory.
